src/rust/engine/process_execution/src/lib.rs

```diff
@@ -391,17 +397,40 @@ pub struct FallibleProcessResultWithPlatform {
 
 /// Metadata for a ProcessResult corresponding to the REAPI `ExecutedActionMetadata` proto. This
 /// conversion is lossy, but the interesting parts are preserved.
-#[derive(Clone, Debug, Default, Eq, PartialEq)]
+#[derive(Clone, Debug, Eq, PartialEq)]
 pub struct ProcessResultMetadata {
   /// The time from starting to completion, including preparing the chroot and cleanup.
   /// Corresponds to `worker_start_timestamp` and `worker_completed_timestamp` from
   /// `ExecutedActionMetadata`.
+  ///
+  /// NB: This is optional because the REAPI does not guarantee that it is returned.
   pub total_elapsed: Option<Duration>,
+  /// The source of the result.
+  pub source: ProcessResultSource,
 }
 
 impl ProcessResultMetadata {
-  pub fn new(total_elapsed: Option<Duration>) -> Self {
-    ProcessResultMetadata { total_elapsed }
+  pub fn new(total_elapsed: Option<Duration>, source: ProcessResultSource) -> Self {
+    ProcessResultMetadata {
+      total_elapsed,
+      source,
+    }
+  }
+
+  pub fn new_from_metadata(metadata: ExecutedActionMetadata, source: ProcessResultSource) -> Self {
```

Comment:
Rust style nit: I've seen some alternate constructors just called `with_foo` or `from_foo` instead of `new_from_foo`. For example: `Vec::with_capacity`. Pure bike shed but would it be better called `with_metadata` or `from_metadata`. Thoughts?

Reply:
We have a few of both styles: I can make a separate pass to unify around `with_`.

Reply:
For my own edification, why prefer this method rather than implementing `From` as before?

Reply:
There are now two required arguments, so `From` would need to be `impl From<(ExecutedActionMetadata, ProcessResultSource)> for ProcessResultMetadata`, which is stretches the API quite a bit.